### DISCUSSION TITLE
Page lifecycle : event when leaving page to trigger function

### DISCUSSION THREAD

**Discussion by stockersky** [zauberzeug/nicegui#2156] [Q&A] [answered]
### Description

First of all, thanks for Nicegui. It looks great !

I am trying to build a dashboard to watch Kubernetes objects.

The Page should initiate some Nicegui Data Elements that will be updated by a background task. When quitting /changing Page, the background task should be terminated.

The following code snippet initiate a Table that get updated by a Kubernetes Watcher when Pods are created or deleted.
This works !
However, when leaving the page, I must terminate the watcher. Otherwise, it will continue to run. In the code below, I achieve it through a hackish Event on the ui.link to another page.

In VueJS, [Component Lifecycle Hooks](https://vuejs.org/guide/essentials/lifecycle#lifecycle-diagram) can trigger methods when a component is created and destroyed.

Most of the Nicegui documentation and examples use Event triggered by the user on Control Widgets, like Button. 
Is there a way to trigger actions when a Page or an Element is created and when it's destroyed (I mean 'destroyed' when its not in the view anymore) ?

```Python
import asyncio
import platform

from nicegui import ui, app, background_tasks
from kubernetes_asyncio import client, config, watch
from icecream import ic

class PodWatcher:
    def __init__(self):
        self.podslist = []
        self.w = watch.Watch()
        self.pods_list()
        
    async def start_watch(self):
        await config.load_kube_config()
        v1 = client.CoreV1Api()
        
        async for event in self.w.stream(v1.list_pod_for_all_namespaces):
            print(f"Event: {event['type']} {event['object'].metadata.name} {event['object'].status.phase}")
            if event['type'] == 'ADDED':
                self.handle_new_pod(
                    {
                        'podname': event['object'].metadata.name,
                        'status': event['object'].status.phase
                    }
                )
            if event['type'] == 'DELETED':
                self.handle_deleted_pod(event['object'].metadata.name)
    
    async def stop_watch(self):
        print('Stoping watcher')
        try:
            self.w.stop()
            await self.w.close()
        except Exception:
            pass

    def handle_new_pod(self, pod: dict) -> None:
        self.podslist.append(pod)
        self.pods_list.refresh()  

    def handle_deleted_pod(self, podname: str) -> None:
        stuff = [ i for i,n in enumerate(podslist) if n['podname'] == podname ]
        ic(stuff)
        for i in stuff: self.podslist.pop(i)
        self.pods_list.refresh()       

    @ui.refreshable
    def pods_list(self) -> None:
        columns = [
            {'name': 'podname', 'label': 'podname', 'field': 'podname', 'required': True, 'align': 'left'},
            {'name': 'status', 'label': 'status', 'field': 'status', 'sortable': True},
        ]
        ui.table(title="pods list", columns=columns, rows=self.podslist, row_key='name')


@ui.page('/')
async def new_name():
    watcher = PodWatcher()
    back = background_tasks.create(watcher.start_watch())
    
    ui.button("Stop watcher", on_click=lambda: watcher.stop_watch())
    ui.link('Visit other page', other_page).on('click', lambda: watcher.stop_watch()) 

@ui.page('/other_page')
def other_page():
    ui.label('Welcome to the other side')

ui.run(reload=platform.system() != "Windows")
```

Well, probably i am misusing Nicegui.

**Comment by rodja** [comment, marked as answer]
Similar to https://nicegui.io/documentation/page#wait_for_client_connection, you can wait for a client to disconnect with 

```py
await client.disconnected()
```

All code below that line will only be executed after the client has left the page. There is also [`app.on_disconnect`](https://nicegui.io/documentation/section_action_events#events) to handle disconnects globally.

**Reply by stockersky** [reply]
Thank you @rodja : it solved this problem 👍
